Doc updates for Aerospike integration (#7510)
solo-io/gloo: docs/content/guides/security/rate_limiting/enterprise/_index.md

weight: 40
---

To enforce rate limiting, Envoy queries an external rate limit server. You can configure changes to the rate limit server as follows.

* [Configure rate limiting query behavior](#query-behavior)
* [Change the default database](#database) from an internal Redis deployment to an external DynamoDB or Aerospike database
* [Debug the rate limit server](#debug)

## Configure Envoy rate limit query behavior {#query-behavior}

To achieve global rate limiting, Envoy queries an external server that is backed by Redis by default. You can configure the default query behavior as described in the following table.

| Query behavior | Default value | How you can change the default | 
| --- | --- | --- |

> | Timeout for the query | 100ms | Change the timeout duration, such as to 200ms.<br><`glooctl edit settings --name default --namespace gloo-system ratelimit --request-timeout=200ms` |

Reviewer: ```suggestion
| Timeout for the query | 100ms | Change the timeout duration, such as to 200ms.<br>`glooctl edit settings --name default --namespace gloo-system ratelimit --request-timeout=200ms` |
```

> | What to do in case of query failure | Allow the request | Deny the request if the rate limit service cannot complete the query.<br>`glooctl edit settings --name default --namespace gloo-system ratelimit --deny-on-failure=true`|

Reviewer: ```suggestion
| Query failure behavior | Allow the request | Deny the request if the rate limit service cannot complete the query.<br>`glooctl edit settings --name default --namespace gloo-system ratelimit --deny-on-failure=true`|
```

## Change the rate limit server's backing database {#database}

> By default, the rate limit server is backed by a Redis instance that Gloo Edge deploys for you. Redis is a good choice for global rate limiting data storage because of its small latency. However, you might want to use a different database for the following reasons:

Reviewer: ```suggestion
By default, the rate limit server is backed by a Redis instance that Gloo Edge deploys for you. Redis is a good choice for global rate limiting data storage because of its low latency. However, you might want to use a different database for the following reasons:
```

* Using an existing database
* Using a database that is external to the cluster, such as for data privacy concerns

Gloo Edge supports the following external databases for the rate limit server:
* [DynamoDB](#dynamodb)
* [Aerospike](#aerospike)

### DynamoDB-backed rate limit server {#dynamodb}

You can use DynamoDB as the backing storage database for the Gloo Edge rate limit server. DynamoDB is built for single-millisecond latencies. It includes features such as built-in replication ([DynamoDB Global Tables](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GlobalTables.html)) that can help you set up global rate limiting across multiple instances or multiple data centers.

{{% notice note %}}
You can use DynamoDB with **Gloo Edge Enterprise** version 0.18.29 or later.
{{% /notice %}}

1. Create a secret in your cluster that includes your AWS credentials for the DynamoDB that you want to use. The secret must be in the same namespace as your Gloo installation, such as `gloo-system`. For more information, see the [AWS docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/SettingUp.DynamoWebService.html).
   ```shell

>    glooctl create secret aws -n gloo-system

Reviewer: should this include `--secret-key <secret> --access-key <key>`?

   ```

> 2. When you [install]({{< versioned_link_path fromRoot="/installation/enterprise/">}}) or [upgrade]({{< versioned_link_path fromRoot="/operations/upgrading/">}}) your Gloo Edge Enterprise Helm installation, complete the following steps:

Reviewer: ```suggestion
2. [Install]({{< versioned_link_path fromRoot="/installation/enterprise/">}}) or [upgrade]({{< versioned_link_path fromRoot="/operations/upgrading/">}}) your Gloo Edge Enterprise Helm installation by completing the following steps:
```

   1. Disable the default Redis server backing storage by setting `rateLimit.enabled` to `false`.
   2. Provide the rate limiting DynamoDB Helm chart configuration options, as shown in the following table.

| Option | Type | Description |
| --- | --- | --- |
| rateLimit.deployment.dynamodb.secretName                  | string   | Required: The name of the secret with the AWS credentials that you previously created. The secret must be in the same namespace as your Gloo installation, such as `gloo-system`.|
| rateLimit.deployment.dynamodb.region                      | string   | The AWS region to run the DynamoDB requests in. The default region is `us-east-2`. |
| rateLimit.deployment.dynamodb.tableName                   | string   | The name of the DynamoDB table that backs the rate limit service. The default name is `rate-limits`. |

> | rateLimit.deployment.dynamodb.consistentReads             | bool     | If `true`, the reading response from DynamoDB is _strongly consistent_, or the most up to date data. The default value is `false`, or _eventually consistent_, which might be less accurate but also has lower latency and less chance of a 500 response than _strongly consistent_. For more information, see the [DynamoDB docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html).|

Reviewer: ```suggestion
| rateLimit.deployment.dynamodb.consistentReads             | bool     | If `true`, the reading response from DynamoDB is _strongly consistent_, or the most up-to-date data. The default value is `false`, or _eventually consistent_, which might be less accurate but also has lower latency and less chance of a 500 response than _strongly consistent_. For more information, see the [DynamoDB docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html).|
```

### Aerospike-backed rate limit server {#aerospike}

You can use [Aerospike](https://docs.aerospike.com/database) as the backing storage database for the Gloo Edge rate limit server. Aerospike is a real-time data platform with support for helpful features such as in-memory storage and streaming.

{{% notice note %}}
You can use Aerospike with **Gloo Edge Enterprise** version 1.13.0 or later.<br><br>
If you use also use Aerospike to store your Gloo Portal API keys, your Aerospike configurations must match. For example, use the same Aerospike IP address, port, and namespace in your Gloo Portal Storage custom resoure configuration and the rate limit server.
{{% /notice %}}

1. Create an Aerospike database instance to use as the backing storage for the rate limit server. For setup steps, see the [Gloo Portal documentation](https://docs.solo.io/gloo-portal/main/guides/portal_features/apikey_storage/). 
2. To rate limit APIs that you manage with Gloo Portal, make sure that your configuration matches the configuration that you used with your [Gloo Portal Storage custom resource](https://docs.solo.io/gloo-portal/main/guides/portal_features/apikey_storage/).

> 3. When you [install]({{< versioned_link_path fromRoot="/installation/enterprise/">}}) or [upgrade]({{< versioned_link_path fromRoot="/operations/upgrading/">}}) your Gloo Edge Enterprise Helm installation, complete the following steps:

Reviewer: ```suggestion
3. [Install]({{< versioned_link_path fromRoot="/installation/enterprise/">}}) or [upgrade]({{< versioned_link_path fromRoot="/operations/upgrading/">}}) your Gloo Edge Enterprise Helm installation by completing the following steps:
```

   1. Disable the default Redis server backing storage by setting `rateLimit.enabled` to `false`.
   2. Provide the rate limiting Aerospike Helm chart configuration options, as shown in the following table. These values match what you configured in your Aerospike database setup. 

| Option | Type | Description |
| --- | --- | --- |

> |rateLimit.deployment.aerospike.address|string|The IP address or hostname of the Aerospike database. The address must be reachable from Gloo Edge, such as in a virtual machine with a public IP address or in a pod in the cluster. When you set this value, you also enable Aerospike database as the backing storage for the rate limit service.|

Reviewer: ```suggestion
|rateLimit.deployment.aerospike.address|string|The IP address or hostname of the Aerospike database. The address must be reachable from Gloo Edge, such as in a virtual machine with a public IP address or in a pod in the cluster. By setting this value, you also enable Aerospike database as the backing storage for the rate limit service.|
```

|rateLimit.deployment.aerospike.namespace|string|The Aerospike namespace of the database. The default value is `solo-namespace`.|
|rateLimit.deployment.aerospike.set|string|The Aerospike name of the database set. The default value is `ratelimiter`.|
|rateLimit.deployment.aerospike.port|int|The port of the `rateLimit.deployment.aerospike.address`. The default port is `3000`.|
|rateLimit.deployment.aerospike.batchSize|int|The size of the batch, which is the number of keys sent in the request. The default value is `5000`.|

> |rateLimit.deployment.aerospike.commitLevel|int|The commit level of guaranteed consistency when commiting a transaction on the Aerospike server. For possible values, see the [Aerospike commit policy](https://github.com/aerospike/aerospike-client-go/blob/master/commit_policy.go). The default value is `1`.|

Reviewer: ```suggestion
|rateLimit.deployment.aerospike.commitLevel|int|The level of guaranteed consistency for transaction commits on the Aerospike server. For possible values, see the [Aerospike commit policy](https://github.com/aerospike/aerospike-client-go/blob/master/commit_policy.go). The default value is `1`.|
```

| Option | Type | Description |
| --- | --- | --- |
| rateLimit.deployment.dynamodb.secretName                  | string   | Required: The name of the secret with the AWS credentials that you previously created. The secret must be in the same namespace as your Gloo installation, such as `gloo-system`.|
| rateLimit.deployment.dynamodb.region                      | string   | The AWS region to run the DynamoDB requests in. The default region is `us-east-2`. |
| rateLimit.deployment.dynamodb.tableName                   | string   | The name of the DynamoDB table that backs the rate limit service. The default name is `rate-limits`. |
| rateLimit.deployment.dynamodb.consistentReads             | bool     | If `true`, the reading response from DynamoDB is _strongly consistent_, or the most up to date data. The default value is `false`, or _eventually consistent_, which might be less accurate but also has lower latency and less chance of a 500 response than _strongly consistent_. For more information, see the [DynamoDB docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html).|

> | rateLimit.deployment.dynamodb.batchSize                   | uint8    | The batch size for `GET` requests to DynamoDB. The max value is `100`, which is also the default value. |

Reviewer 2: I believe this is the number of requests.  This is set to the default value of `5000`.

> | rateLimit.deployment.dynamodb.batchSize                   | uint8    | The batch size for `GET` requests to DynamoDB. The max value is `100`, which is also the default value. |

Author: @jackstine thanks...how can the default value be 5000 if the max value is 100?

As part of the rate limit service deployment, Gloo Edge uses the provided AWS credentials to automatically create the rate limits DynamoDB table (default name `rate-limits`) in your AWS region (default `us-east-2`). If you want to turn the table into a globally replicated table, you
can select the regions to replicate to in the DynamoDB AWS console UI.

### Aerospike-backed rate limit server {#aerospike}

> You can use [Aerospike](https://docs.aerospike.com/database) as the backing storage database for the Gloo Edge rate limit server. Aerospike is a real-time data platform with support for helpful features such as in-memory storage and streaming.

Author: Do they _have_ to BYO aerospike database? or is it an option, and if they don't, we'll create one for them? or...?

{{% notice note %}}
You can use Aerospike with **Gloo Edge Enterprise** version 1.13.0 or later.<br><br>
If you use also use Aerospike to store your Gloo Portal API keys, your Aerospike configurations must match. For example, use the same Aerospike IP address, port, and namespace in your Gloo Portal Storage custom resoure configuration and the rate limit server.
{{% /notice %}}

1. Create a secret in your cluster that includes your AWS credentials for the DynamoDB that you want to use. The secret must be in the same namespace as your Gloo installation, such as `gloo-system`. For more information, see the [AWS docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/SettingUp.DynamoWebService.html).
   ```shell
   glooctl create secret aws -n gloo-system
   ```

> 2. When you [install]({{< versioned_link_path fromRoot="/installation/enterprise/">}}) or [upgrade]({{< versioned_link_path fromRoot="/operations/upgrading/">}}) your Gloo Edge Enterprise Helm installation, complete the following steps:

Author: I wanted to confirm, can you add Aerospike to the RL config during an upgrade to the Helm chart, or only at installation?

Reviewer 2: Maybe someone else can answer this?  I am not sure.

Reviewer 3: You should be able to add the config during an upgrade, but I'm don't think this has been tested.

   1. Disable the default Redis server backing storage by setting `rateLimit.enabled` to `false`.

>    2. Provide the rate limiting Aerospike Helm chart configuration options, as shown in the following table.

Author: General question for the Helm chart settings, do they have default values? I didn't notice many in the [descriptions](https://github.com/solo-io/solo-projects/blob/master/install/helm/gloo-ee/reference/values.txt), only for a couple.

| Option | Type | Description |
| --- | --- | --- |
|rateLimit.deployment.aerospike.address|string|The IP address or hostname of the Aerospike database. The address must be reachable from Gloo Edge, such as in a virtual machine with a public IP address or in a pod in the cluster. When you set this value, you also enable Aerospike database as the backing storage for the rate limit service.|
|rateLimit.deployment.aerospike.namespace|string|The Aerospike namespace of the database. Defaults to `solo-namespace`.|
|rateLimit.deployment.aerospike.set|string|The Aerospike name of the database set. Defaults to `ratelimiter`.|
|rateLimit.deployment.aerospike.port|int|The port of the `rateLimit.deployment.aerospike.address`. The default port is `3000`.|

> |rateLimit.deployment.aerospike.batchSize|int|The size of the batch.|

Author: For the batch size,
* what is this unit for this batch size value, MB?
* what is the max size? 100 like for DynamoDB?

Reviewer 2: the unit is number of requests.
I am not sure what the max size is. I believe this is set with the [this parameter in the Aerospike configuration](https://docs.aerospike.com/server/operations/manage/manage_batches#:~:text=Description-,batch%2Dmax%2Drequests,-5000) which we do not control on the client side.

|rateLimit.deployment.aerospike.commitLevel|int|The commit level of guaranteed consistency when commiting a transaction on the Aerospike server. For possible values, see the [Aerospike commit policy](https://github.com/aerospike/aerospike-client-go/blob/master/commit_policy.go).|
|rateLimit.deployment.aerospike.readModeSC|int|The read mode for strong consistency (SC) options. For possible values, see the [Aerospike read mode SC](https://github.com/aerospike/aerospike-client-go/blob/master/read_mode_sc.go).|
|rateLimit.deployment.aerospike.readModeAP|int|The read mode for availability (AP). For possible values, see the [Aerospike read mode AP](https://github.com/aerospike/aerospike-client-go/blob/master/read_mode_ap.go).|

> |rateLimit.deployment.aerospike.tls.name|string|The TLS server name. For more information, see [Aerospike TLS Name Clarification](https://docs.aerospike.com/server/guide/security/tls#tls-name-clarification).|

Author: what is the TLS name? is it the server or client name (ref Aerospike [doc](https://docs.aerospike.com/server/guide/security/tls#tls-name-clarification))?

Reviewer 2: [Here is more information](https://docs.aerospike.com/reference/configuration#tls-name).  The TLS name is the subject name of the TLS authority.

Author: ok thanks, I will update it to that doc!

|rateLimit.deployment.aerospike.tls.version|string|The TLS version. Version 1.0, 1.1, 1.2, and 1.3 are supported.|

> |rateLimit.deployment.aerospike.tls.insecure|bool|The TLS insecure setting.|

Author: what does this TLS insecure mean? What is the behavior if true? What if false?

Reviewer 2: If set, do not authenticate the authority of the certificate on the server's end. Generally only used in non-prod environments for testing, when the certificate is not known.

|rateLimit.deployment.aerospike.tls.certSecretName|string| The name of the `kubernetes.io/tls` secret that has the `tls.crt` and `tls.key` data.|
|rateLimit.deployment.aerospike.tls.rootCASecretName|string|The secret name for the Opaque root CA that sets the key as `tls.crt`.|

> |rateLimit.deployment.aerospike.tls.curveGroups[]|string|The TLS curve groups.|

Author: what is this TLS curve group, like what would example values look like?

Reviewer 2: I do not fully understand TLS curve groups, but this is an option. 
CurveID is the type of a TLS identifier for an elliptic curve. See https://www.iana.org/assignments/tls-parameters/tls-parameters.xml#tls-parameters-8.
